packages/react/src/utilities/hooks/useCheckboxGroup/useCheckboxGroup.ts

```diff
@@ -153,18 +163,19 @@ export function useCheckboxGroup({
       };
 
       useEffect(() => {
-        if (!allowIndeterminate || !inputRef.current) return;
-        toggleIndeterminate(inputRef.current, getInputs);
-        /* We can use `currentValue` as dependency here,
-      since that is omly when this checkbox changes */
+        if (!allowIndeterminate) return;
+
+        toggleIndeterminate(getIndeterminateInputs, getInputs);
       }, [groupValue]);
 
       useEffect(() => {
-        if (inputRef.current && !allowIndeterminate) {
-          checkboxRefs.current.set(value, inputRef.current);
-        }
+        if (!inputRef.current) return;
+
+        const refs = allowIndeterminate ? indeterminateRefs : checkboxRefs;
+        refs.current.add(inputRef.current);
+
         return () => {
-          checkboxRefs.current.delete(value);
+          if (inputRef.current) refs.current.delete(inputRef.current);
```

Comment:
Oh just remembered; the React hooks eslint points out that any `ref.current` _might_ change before `return` runs, so to be 100% robust, we can do:
```
 useEffect(() => {
        if (!inputRef.current) return;
        const input = inputRef.current;
        const refs = allowIndeterminate ? indeterminateRefs : checkboxRefs;
        refs.current.add(input);

        return () => input.delete(inputRef.current);
}, [value]);
```